XMonad/Actions/Submap.hs

```diff
@@ -71,28 +117,38 @@ submapDefault = submapDefaultWithKey . const
 submapDefaultWithKey :: ((KeyMask, KeySym) -> X ())
                      -> M.Map (KeyMask, KeySym) (X ())
                      -> X ()
-submapDefaultWithKey defAction keys = do
-    XConf { theRoot = root, display = d } <- ask
-
-    io $ grabKeyboard d root False grabModeAsync grabModeAsync currentTime
-    io $ grabPointer d root False buttonPressMask grabModeAsync grabModeAsync
-                     none none currentTime
-
-    (m, s) <- io $ allocaXEvent $ \p -> fix $ \nextkey -> do
-        maskEvent d (keyPressMask .|. buttonPressMask) p
-        ev <- getEvent p
-        case ev of
-          KeyEvent { ev_keycode = code, ev_state = m } -> do
-            keysym <- keycodeToKeysym d code 0
-            if isModifierKey keysym
-                then nextkey
-                else return (m, keysym)
-          _ -> return (0, 0)
-    -- Remove num lock mask and Xkb group state bits
-    m' <- cleanMask $ m .&. ((1 `shiftL` 12) - 1)
-
-    io $ ungrabPointer d currentTime
-    io $ ungrabKeyboard d currentTime
-    io $ sync d False
-
-    fromMaybe (defAction (m', s)) (M.lookup (m', s) keys)
+submapDefaultWithKey defAction keys = submapWorker id $
+    \(m', s) -> fromMaybe (defAction (m', s)) (M.lookup (m', s) keys)
+
+-----------------------------------------------------------------------
+-- Internal stuff
+
+submapWorker :: (X (KeyMask, KeySym) -> X (KeyMask, KeySym)) -- ^ Get keys
+             -> ((KeyMask, KeySym) -> X ())                  -- ^ Lookup and execute action, if any
+             -> X ()
+submapWorker getKeys lookupAction = do
+    XConf{ theRoot = root, display = dpy } <- ask
+
+    io $ do grabKeyboard dpy root False grabModeAsync grabModeAsync currentTime
+            grabPointer dpy root False buttonPressMask grabModeAsync grabModeAsync
+                        none none currentTime
+
+    (m', s) <- getKeys $ do
+        (m, s) <- io $ allocaXEvent $ \p -> fix $ \nextkey -> do
+            maskEvent dpy (keyPressMask .|. buttonPressMask) p
+            ev <- getEvent p
+            case ev of
+              KeyEvent { ev_keycode = code, ev_state = m } -> do
+                keysym <- keycodeToKeysym dpy code 0
+                if isModifierKey keysym
+                    then nextkey
+                    else return (m, keysym)
+              _ -> return (0, 0)
+        -- Remove num lock mask and Xkb group state bits
+        m' <- cleanMask $ m .&. ((1 `shiftL` 12) - 1)
+        io $ do ungrabPointer dpy currentTime
+                ungrabKeyboard dpy currentTime
+                sync dpy False
+        pure (m', s)
+
+    lookupAction (m', s)
```

Comment:
We could totally do that (I didn't put too much thought into how ghastly the type signature of `submapWorker` looks because it's not exported anyways).  The new ergonomics also don't seem too bad: 

``` haskell
    withSimpleWindow wc descriptions submapWorker >>= \(m', s) ->
        maybe (pure ()) snd (M.lookup (m', s) keys)
```